design_documents/sncast_interface_overhaul.md

```diff
@@ -0,0 +1,47 @@
+# sncast interface overhaul
+
+## Context
+In light of needed changes on how the scripts and contract declaration work (ie adding support for scarb workspaces and various directory layouts),
+we need to re-think how we call the respective subcommands, and try to improve it. This design doc covers only a portion
+(although significant one) of sncast interface - it was decided recently we should make the interfaces as 'clap-native'
+as possible, but this design doc won't cover those changes in detail, as they are pretty straight forward and already have
+an opened issue.
+
+## The 'now'
+Currently, when declaring a contract we require user to pass the contract name using `--contract-name` flag. This flag only
+takes the contract's name, so for example it is not possible to specify the package this contract was defined in (only
+achievable when passing `--path-to-scarb-toml`). 
+
+When running a script we require a positional argument, that is the script package name. Similarly to declare, we could
+pass `--path-to-scarb-toml` if we wanted to pick the manifest we wanted to use.
+
+These approaches are a bit different to each other, and have a common flaw in the shape of having to specify a manifest 
+file one wants to work with, when the package needs to be specified. This may seem quite quirky, as we probably could 
+assume that we're always in some workspace/package.
+
+## Proposed changes
+We could streamline the way we interact with scarb-related subcommands (declare, script at the moment of writing this)
+to look like this:
+
+```bash
+$ # this is the same module
+$ sncast script package_namea::path::to::script_module
+$ sncast script to::script_module
+$ sncast script script_module
+
+$ # this is the same contract
+$ sncast declare package_namea::path::to::contract
+$ sncast to::contract
+$ sncast script script_module
+```
+
+In case there is only one script module/contract with specified name in workspace, only this name can be used to target it.
+Otherwise, if it is ambiguous in the context of workspace, we should show all matches with full paths, to let user decide
```

Comment:
Isn't it better to explicitly provide the `--package` flag and then specify the path as if it were from the source root of the package?
```
$ sncast script --package my_package first_mod::second_mod::script_mod
```
This way we always know where to look for, there is no ambiguity allowed in the first place

Reply:
What you suggested is basically the same idea to what's in this design doc, only that I move the package to the script path and make it non required (if fiven mod is unique). I'm not sure how is it better to have additional, required flag tbh.

Reply:
Okay, I had a wrong assumption that we want to support workspaces for other commands, mb. Tbh I prefer this `--package` arg to unify the interface with my vision of `sncast declare` (see above)

Reply:
I think @piotmag769 is right here, it would be better to move package name under the package flag. It would result in interface consistent with scarb, cargo and snforge.

In rust when you want to run one test you specify
```
 cargo test --package forge --test main -- integration::env::read_short_string --exact --nocapture 
```

So if there is no flag we should search for path in all packages. If there is a flag we should search for it in specified packages. If there are two results we should throw an ambiguity error.